pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3434 @@
+        _stack = [ (node, node._args_, 0, node.nargs(), [False])]
+        #
+        # Iterate until the stack is empty
+        #
+        # Note: 1 is faster than True for Python 2.x
+        #
+        while 1:
+            #
+            # Get the top of the stack
+            #   _obj        Current expression object
+            #   _argList    The arguments for this expression objet
+            #   _idx        The current argument being considered
+            #   _len        The number of arguments
+            #   _result     The 'dirty' flag followed by return values
+            #
+            _obj, _argList, _idx, _len, _result = _stack.pop()
+            #
+            # Iterate through the arguments
+            #
+            while _idx < _len:
+                _sub = _argList[_idx]
+                _idx += 1
+                flag, value = self.visiting_potential_leaf(_sub)
+                if flag:
+                    if id(value) != id(_sub):
+                        _result[0] = True
+                    _result.append( value )
+                else:
+                    #
+                    # Push an expression onto the stack
+                    #
+                    _stack.append( (_obj, _argList, _idx, _len, _result) )
+                    _obj                    = _sub
+                    #_argList                = self.children(_sub)
+                    _argList                = _sub._args_
+                    _idx                    = 0
+                    _len                    = _sub.nargs()
+                    _result                 = [False]
+            #
+            # Process the current node
+            #
+            # If the user has defined a visit() function in a
+            # subclass, then call that function.  But if the user
+            # hasn't created a new class and we need to, then
+            # call the ExpressionReplacementVisitor.visit() function.
+            #
+            ans = self.visit(_obj, _result[1:])
+            if _result[0] and id(ans) == id(_obj):
+                ans = self.construct_node(_obj, _result[1:])
+            if _result[0] and ans.__class__ is MonomialTermExpression:
+                ans.__class__ = ProductExpression
+            if _stack:
+                if _result[0]:
+                    _stack[-1][-1][0] = True
+                #
+                # "return" the recursion by putting the return value on the end of the results stack
+                #
+                _stack[-1][-1].append( ans )
+            else:
+                return self.finalize(ans)
+
+
+
+#-------------------------------------------------------
+#
+# Functions used to process expression trees
+#
+#-------------------------------------------------------
+
+# =====================================================
+#  clone_expression
+# =====================================================
+
+class _CloneVisitor(ExpressionValueVisitor):
```

Comment:
I do not like using a visitor to implement cloning.  If the point is to avoid deepcopy it won't work, as deepcopying a model (or block) will end up copying (cloning) the expressions in the model using deepcopy and not this method.  That leads to having two ways of duplicating expressions depending on the execution path -- and that strikes me as asking for trouble.

Reply:
The semantics for deepcopy() and clone_expression() are different, which justifies a separate function.